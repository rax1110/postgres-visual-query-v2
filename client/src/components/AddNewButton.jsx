import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';
import * as PropTypes from 'prop-types';
import { Button } from 'reactstrap';

export const AddNewButton = ({ id, size, onClick }) => (
  <Button
    className=""
    outline
    color="info"
    size={size}
    id={id}
    onClick={onClick}
  >
    <FontAwesomeIcon icon="plus" />
  </Button>
);

AddNewButton.propTypes = {
  id: PropTypes.string,
  size: PropTypes.string,
  onClick: PropTypes.func,
};
