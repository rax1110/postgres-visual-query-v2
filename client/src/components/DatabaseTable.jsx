import React from 'react';
import { Button, Tooltip } from 'reactstrap';
import { connect } from 'react-redux';
import _ from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { addColumn, addTable, removeTable } from '../actions/queryActions';
import { withToggle } from '../hocs/withToggle';

export const DatabaseTable = ({ data, checked, id, addTableProp, toggle, toggleStatus }) => {
  const handleOnClick = () => {
    addTableProp(data);
  };

  let tableTypeColor = 'primary';

  if (_.isEqual(data.table_type, 'VIEW')) {
    tableTypeColor = 'info';
  } else if (_.isEqual(data.table_type, 'FOREIGN')) {
    tableTypeColor = 'secondary';
  } else if (_.isEqual(data.table_type, 'MATERIALIZED VIEW')) {
    tableTypeColor = 'light';
  }

  const btnSelected = checked ? 'success' : tableTypeColor;

  const modifiers = {
    preventOverflow: {
      enabled: false,
    },
    hide: {
      enabled: false,
    },
  };

  return (
    <div className="w-100 pr-1">
      <Button
        size="sm"
        color={btnSelected}
        id={id}
        className="btn-block my-1 pt-0 text-left"
        onClick={handleOnClick}
      >
        <small
          color={tableTypeColor}
          className="text-truncate align-self-start"
        >
          <span className="mr-1 px-0">
            <FontAwesomeIcon icon="th-large" />
          </span>
          {data.table_type}
        </small>
        <div className="text-truncate">
          {` ${data.table_name}`}
        </div>
      </Button>
      <Tooltip
        placement="right"
        isOpen={toggleStatus}
        target={id}
        toggle={toggle}
        modifiers={modifiers}
        delay={{ show: 200, hide: 0 }}
        className=""
      >
        {data.table_name}
      </Tooltip>
    </div>
  );
};

DatabaseTable.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  checked: PropTypes.bool,
  id: PropTypes.string,
  addTableProp: PropTypes.func,
  toggle: PropTypes.func,
  toggleStatus: PropTypes.bool,
};

const mapDispatchToProps = {
  addColumn,
  addTableProp: data => addTable(data),
  removeTable,
};

export default withToggle(connect(null, mapDispatchToProps)(DatabaseTable));
