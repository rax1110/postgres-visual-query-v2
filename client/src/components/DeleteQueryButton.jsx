import React from 'react';
import { Button } from 'reactstrap';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { deleteQuery, setActiveQuery } from '../actions/queryActions';
import { deleteQueries, removeMainFromQueries } from '../actions/queriesActions';

export const DeleteQueryButton = (props) => {
  const handleOnClick = () => {
    props.deleteQuery();

    const mainQuery = props.queries.find(query => query.id === 0);

    if (mainQuery) {
      props.removeMainFromQueries();
      props.setActiveQuery(mainQuery);
    } else {
      props.deleteQueries();
    }
  };

  return (
    <Button color="danger" className="mr-2" onClick={handleOnClick}>
      <FontAwesomeIcon icon="trash" />
    </Button>
  );
};

DeleteQueryButton.propTypes = {
  deleteQuery: PropTypes.func,
  deleteQueries: PropTypes.func,
  queries: PropTypes.arrayOf(PropTypes.shape({})),
  removeMainFromQueries: PropTypes.func,
  setActiveQuery: PropTypes.func,
};

const mapStateToProps = store => ({
  queries: store.queries,
});

const mapDispatchToProps = {
  deleteQuery,
  deleteQueries,
  removeMainFromQueries,
  setActiveQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(DeleteQueryButton);
