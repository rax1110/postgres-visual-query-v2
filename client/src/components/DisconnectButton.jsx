import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Redirect } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import { disconnect } from '../actions/hostActions';
import { translations } from '../utils/translations';

export const DisconnectButton = (props) => {
  const handleOnClick = () => {
    props.disconnect();
  };

  if (!props.connected) {
    return <Redirect to="/" />;
  }

  return (
    <Button size="lg" className="btn-block my-2 px-2" color="danger" onClick={handleOnClick}>
      {translations[props.language.code].sideBar.disconnectB}
    </Button>
  );
};

DisconnectButton.propTypes = {
  disconnect: PropTypes.func,
  connected: PropTypes.bool,
  language: PropTypes.shape({ code: PropTypes.string }),
};

const mapStateToProps = store => ({
  connected: store.host.connected,
  language: store.settings.language,
});

const mapDispatchToProps = {
  disconnect,
};

export default connect(mapStateToProps, mapDispatchToProps)(DisconnectButton);
