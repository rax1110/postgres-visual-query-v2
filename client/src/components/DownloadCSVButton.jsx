import React from 'react';
import { connect } from 'react-redux';
import { CSVLink } from 'react-csv';
import _ from 'lodash';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadCSVButton = (props) => {
  const getHeaders = (result) => {
    if (_.isNull(result)) {
      return [];
    }

    const headers = [];

    result.fields.forEach((field) => {
      const header = {
        label: field.name,
        key: field.name,
      };

      headers.push(header);
    });

    return headers;
  };

  const getData = (result) => {
    if (_.isNull(result)) {
      return [];
    }
    return result.rows;
  };

  const disabled = _.isNull(props.result);

  return (
    <CSVLink
      className="mr-2"
      data={getData(props.result)}
      headers={getHeaders(props.result)}
      filename="result.csv"
    >
      <Button disabled={disabled}>
        <FontAwesomeIcon icon="download" />
        <div className="d-inline"> CSV</div>
      </Button>
    </CSVLink>
  );
};

DownloadCSVButton.propTypes = {
  result: PropTypes.shape({}),
};

const mapStateToProps = store => ({
  result: store.query.result,
});

export default connect(mapStateToProps)(DownloadCSVButton);
