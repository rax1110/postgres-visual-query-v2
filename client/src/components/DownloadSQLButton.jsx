import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadSQLButton = (props) => {
  const downloadContent = (name, content) => {
    if (navigator.msSaveBlob) {
      const blobObject = new Blob([content], { type: 'text/plain' });

      window.navigator.msSaveOrOpenBlob(blobObject, name);
    } else {
      const aTag = document.createElement('a');
      const file = new Blob([content], { type: 'text/plain' });

      aTag.href = URL.createObjectURL(file);
      aTag.download = name;

      document.body.appendChild(aTag);
      aTag.click();
    }
  };

  return (
    <Button
      className="mr-2"
      onClick={() => downloadContent('select_query.sql', props.sql)}
    >
      <FontAwesomeIcon icon="download" />
      <div className="d-inline"> SQL</div>
    </Button>
  );
};

DownloadSQLButton.propTypes = {
  sql: PropTypes.string,
};

const mapStateToProps = store => ({
  sql: store.query.sql,
});

export default connect(mapStateToProps)(DownloadSQLButton);
