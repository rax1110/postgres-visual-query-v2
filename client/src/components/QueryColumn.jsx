import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Button,
  ButtonDropdown,
  Card,
  CardBody,
  Container,
  CustomInput,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  Form,
  FormGroup,
  Input,
  InputGroup,
  InputGroupAddon,
  Row,
  UncontrolledTooltip,
} from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Draggable } from 'react-beautiful-dnd';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import { addColumn, removeColumn, updateColumn } from '../actions/queryActions';
import { translations } from '../utils/translations';
import { bannedWords } from '../utils/bannedWords';
import { getCorrectQueryName } from '../utils/getCorrectQueryName';

const CopyButton = ({ id, handleCopy, languageCode }) => (
  <div>
    <Button
      size="sm"
      color="secondary"
      id={`${id}_copy`}
      className="mr-1"
      onClick={handleCopy}
    >
      <FontAwesomeIcon icon="copy" />
    </Button>
    <UncontrolledTooltip placement="top" target={`${id}_copy`} delay={{ show: 500, hide: 0 }}>
      {translations[languageCode].tooltips.copyColumn}
    </UncontrolledTooltip>
  </div>
);

CopyButton.propTypes = {
  id: PropTypes.string,
  handleCopy: PropTypes.func,
  languageCode: PropTypes.string,
};

const RemoveButton = ({ id, handleRemoveColumn }) => (
  <div>
    <Button
      size="sm"
      color="danger"
      id={`${id}_remove`}
      onClick={handleRemoveColumn}
    >
      <FontAwesomeIcon icon="times" />
    </Button>
  </div>
);

RemoveButton.propTypes = {
  id: PropTypes.string,
  handleRemoveColumn: PropTypes.func,
};

export class QueryColumn extends Component {
  constructor(props) {
    super(props);

    this.state = {
      column_alias: props.data.column_alias,
      column_filter: props.data.column_filter,
      filter_valid: true,
      dropDownOpen: false,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
    this.handleSave = this.handleSave.bind(this);
    this.handleSwitch = this.handleSwitch.bind(this);
    this.handleRemoveColumn = this.handleRemoveColumn.bind(this);
    this.handleCopy = this.handleCopy.bind(this);
    this.handleDropDown = this.handleDropDown.bind(this);
  }

  componentDidMount() {
    if (this.props.data.subqueryId) {
      let column = _.cloneDeep(this.props.data);

      const subquerySql = this.props.queries
        .find(query => query.id === this.props.data.subqueryId).sql;

      column = {
        ...column,
        subquerySql,
      };

      this.props.updateColumn(column);
    }
  }

  handleDropDown() {
    this.setState(prevState => ({
      dropDownOpen: !prevState.dropDownOpen,
    }));
  }

  handleRemoveColumn() {
    this.props.removeColumn(this.props.data);
  }

  handleChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleCopy() {
    this.props.addColumn(this.props.data);
  }

  handleRemove(e) {
    const { state } = this;

    this.setState({
      ...state,
      [e.currentTarget.name]: '',
    });

    let column = _.cloneDeep(this.props.data);

    column = {
      ...column,
      [e.currentTarget.name]: '',
    };

    if (_.isEqual(e.target.name, 'column_filter')) {
      this.setState({
        filter_valid: true,
      });
    }

    this.props.updateColumn(column);
  }

  handleSave(e) {
    let column = _.cloneDeep(this.props.data);

    if (e.currentTarget.name === 'subqueryDefault') {
      column = {
        ...column,
        subqueryId: 0,
        subquerySql: '',
      };
    }

    if (e.currentTarget.name === 'subqueryId') {
      const subqueryId = +e.target.value;
      const subquerySql = this.props.queries.find(query => query.id === subqueryId).sql;

      column = {
        ...column,
        subqueryId,
        subquerySql,
      };
    } else {
      column = {
        ...column,
        [e.target.name]: e.target.value,
      };
    }

    let contains = false;

    const filter = _.lowerCase(column.column_filter).split(' ');

    bannedWords.forEach((el) => {
      if (filter.includes(el)) {
        contains = true;
      }
    });

    if (contains) {
      this.setState({
        filter_valid: false,
      });
    } else {
      this.setState({
        filter_valid: true,
      });

      this.props.updateColumn(column);
    }
  }

  handleSwitch(e) {
    let column = _.cloneDeep(this.props.data);

    column = {
      ...column,
      [e.target.name]: !column[e.target.name],
    };

    this.props.updateColumn(column);
  }

  render() {
    const orderDirection = this.props.data.column_order_dir
      ? translations[this.props.language.code].queryBuilder.ascL
      : translations[this.props.language.code].queryBuilder.descL;
    const columnOrderVisibility = this.props.data.column_order ? 'visible' : 'invisible';
    const columnName = _.isEmpty(this.props.data.table_alias)
      ? `${this.props.data.table_name}.${this.props.data.column_name}`
      : `${this.props.data.table_alias}.${this.props.data.column_name}`;
    const filterValid = this.state.filter_valid ? '' : 'is-invalid';
    const linkedQuery = this.props.queries.find(query => query.id === this.props.data.subqueryId);
    const linkedQueryName = linkedQuery
      ? getCorrectQueryName(this.props.language, linkedQuery.queryName, linkedQuery.id)
      : translations[this.props.language.code].queryBuilder.linkSq;

    return (
      <Draggable
        draggableId={`${this.props.id}`}
        index={this.props.index}
      >
        {provided => (
          <div className="m-auto">
            <Card
              className="px-0 my-2"
              {...provided.draggableProps}
              {...provided.dragHandleProps}
              innerRef={provided.innerRef}
            >
              <CardBody className="mx-0 pr-2 pl-1 pt-1 pb-1">
                <Form inline className="align-content-center">
                  <Container fluid className="pr-0">
                    <Row form>
                      <div className="col-auto pl-0 pr-3 d-flex align-items-center">
                        <FontAwesomeIcon className="" icon="sort" />
                      </div>
                      <div className="col-10 p-0">
                        <Row form>
                          <div className="col-auto d-flex">
                            <CustomInput
                              className=""
                              type="checkbox"
                              id={`display-${this.props.data.id}`}
                              name="display_in_query"
                              checked={this.props.data.display_in_query}
                              onChange={this.handleSwitch}
                            />
                            <small
                              className="mr-2 align-self-center text-muted"
                            >
                              {`${this.props.data.table_schema}`}
                            </small>
                            <h6
                              className="m-0 mr-2 align-self-center"
                              id="column_name"
                            >
                              {columnName}
                            </h6>
                            <UncontrolledTooltip
                              placement="top"
                              delay={{ show: 0, hide: 0 }}
                              target="column_name"
                            >
                              {this.props.data.data_type}
                            </UncontrolledTooltip>
                          </div>
                          <div className="col-auto">
                            <FormGroup>
                              <CustomInput
                                className="mr-2"
                                disabled={this.props.distinct}
                                type="switch"
                                id={`column-distinct-on-${this.props.data.id}`}
                                name="column_distinct_on"
                                checked={this.props.data.column_distinct_on}
                                onChange={this.handleSwitch}
                                label={
                                  translations[this.props.language.code].queryBuilder.distinctOnL
                                }
                              />
                              <CustomInput
                                className="mr-2"
                                type="switch"
                                id={`column-group-by-${this.props.data.id}`}
                                name="column_group_by"
                                checked={this.props.data.column_group_by}
                                onChange={this.handleSwitch}
                                label={translations[this.props.language.code].queryBuilder.groupByL}
                              />
                              <CustomInput
                                className="mr-2"
                                type="switch"
                                id={`column-order-${this.props.data.id}`}
                                name="column_order"
                                checked={this.props.data.column_order}
                                onChange={this.handleSwitch}
                                label={translations[this.props.language.code].queryBuilder.orderL}
                              />
                              <CustomInput
                                className={columnOrderVisibility}
                                type="switch"
                                id={`column-order-dir-${this.props.data.id}`}
                                name="column_order_dir"
                                checked={this.props.data.column_order_dir}
                                onChange={this.handleSwitch}
                                label={orderDirection}
                              />
                            </FormGroup>
                          </div>
                        </Row>
                        <Row form>
                          <div className="col-auto">
                            <CustomInput
                              bsSize="sm"
                              type="select"
                              name="column_aggregate"
                              id={`column-aggregate-${this.props.data.id}`}
                              className="my-1 align-self-start"
                              value={this.props.data.column_aggregate}
                              onChange={this.handleSave}
                            >
                              <option value="">
                                {translations[this.props.language.code].queryBuilder.selectFunction}
                              </option>
                              <option value="AVG">AVG</option>
                              <option value="BIT_AND">BIT_AND</option>
                              <option value="BIT_OR">BIT_OR</option>
                              <option value="BOOL_AND">BOOL_AND</option>
                              <option value="BOOL_OR">BOOL_OR</option>
                              <option value="COUNT">COUNT</option>
                              <option value="MAX">MAX</option>
                              <option value="MIN">MIN</option>
                              <option value="SUM">SUM</option>
                              <option value="ASCII">ASCII</option>
                              <option value="BIT_LENGTH">BIT_LENGTH</option>
                              <option value="CHAR_LENGTH">CHAR_LENGTH</option>
                              <option value="INITCAP">INITCAP</option>
                              <option value="LENGTH">LENGTH</option>
                              <option value="LOWER">LOWER</option>
                              <option value="OCTET_LENGTH">OCTET_LENGTH</option>
                              <option value="REVERSE">REVERSE</option>
                              <option value="UPPER">UPPER</option>
                              <option value="TO_ASCII">TO_ASCII</option>
                              <option value="TO_HEX">TO_HEX</option>
                            </CustomInput>
                          </div>
                          <div className="col-auto">
                            <InputGroup
                              className="my-1 align-self-start"
                              size="sm"
                            >
                              <Input
                                style={{ flexBasis: 'auto' }}
                                className="text-dark"
                                type="text"
                                name="column_alias"
                                id={`column-alias-${this.props.data.id}`}
                                onBlur={this.handleSave}
                                onChange={this.handleChange}
                                value={this.state.column_alias}
                                placeholder={
                                  translations[this.props.language.code].queryBuilder.aliasPh
                                }
                              />
                              <UncontrolledTooltip
                                placement="top"
                                delay={{ show: 500, hide: 0 }}
                                target={`column-alias-${this.props.data.id}`}
                              >
                                {translations[this.props.language.code].tooltips.columnAlias}
                              </UncontrolledTooltip>
                              <InputGroupAddon addonType="append">
                                <Button
                                  color="danger"
                                  id={`column-alias-btn-${this.props.data.id}`}
                                  name="column_alias"
                                  onClick={this.handleRemove}
                                >
                                  <FontAwesomeIcon icon="times" />
                                </Button>
                              </InputGroupAddon>
                            </InputGroup>
                          </div>
                          <div className="col-auto" style={{ minWidth: '35%' }}>
                            <InputGroup className="my-1 align-self-start" size="sm">
                              <Input
                                type="text-dark"
                                name="column_filter"
                                id={`column-filter-${this.props.data.id}`}
                                className={filterValid}
                                onBlur={this.handleSave}
                                onChange={this.handleChange}
                                value={this.state.column_filter}
                                placeholder={
                                  translations[this.props.language.code].queryBuilder.filterPh
                                }
                              />
                              <div className="invalid-feedback order-1">
                                {translations[this.props.language.code].tooltips.invalidFilter}
                              </div>
                              <UncontrolledTooltip
                                placement="top"
                                delay={{ show: 500, hide: 0 }}
                                target={`column-filter-${this.props.data.id}`}
                              >
                                {translations[this.props.language.code].tooltips.columnFilter}
                              </UncontrolledTooltip>
                              <InputGroupAddon addonType="append">
                                <ButtonDropdown
                                  isOpen={this.state.dropDownOpen}
                                  toggle={this.handleDropDown}
                                  id={`link-subquery-${this.props.data.id}`}
                                >
                                  <DropdownToggle
                                    className="btn-sm btn-light btn-outline-secondary"
                                    style={{ borderColor: '#d3d8de' }}
                                    caret
                                  >
                                    {linkedQueryName}
                                  </DropdownToggle>
                                  <DropdownMenu>
                                    <DropdownItem
                                      key="query-link-SQ"
                                      id={`subquery-default-${this.props.data.id}`}
                                      name="subqueryDefault"
                                      value=""
                                      onClick={this.handleSave}
                                    >
                                      {translations[this.props.language.code].queryBuilder.linkSq}
                                    </DropdownItem>
                                    {this.props.queries.map((query, index) => (
                                      <DropdownItem
                                        key={`query-${query.id}-column-${this.props.data.id}`}
                                        id={`subquerySql-${index}-${this.props.data.id}`}
                                        name="subqueryId"
                                        value={query.id}
                                        onClick={this.handleSave}
                                      >
                                        {getCorrectQueryName(
                                          this.props.language, query.queryName, query.id,
                                        )}
                                      </DropdownItem>
                                    ))}
                                  </DropdownMenu>
                                </ButtonDropdown>
                                <UncontrolledTooltip
                                  placement="top"
                                  delay={{ show: 500, hide: 0 }}
                                  target={`link-subquery-${this.props.data.id}`}
                                >
                                  {translations[this.props.language.code].tooltips.linkSq}
                                </UncontrolledTooltip>
                              </InputGroupAddon>
                              <InputGroupAddon addonType="append">
                                <Button
                                  color="danger"
                                  id={`column-filter-btn-${this.props.data.id}`}
                                  name="column_filter"
                                  onClick={this.handleRemove}
                                >
                                  <FontAwesomeIcon icon="times" />
                                </Button>
                              </InputGroupAddon>
                            </InputGroup>
                          </div>
                          <div className="col-auto pt-2">
                            <CustomInput
                              className=""
                              type="switch"
                              id={`filter-as-having-${this.props.data.id}`}
                              name="filter_as_having"
                              checked={this.props.data.filter_as_having}
                              onChange={this.handleSwitch}
                              label={translations[this.props.language.code].queryBuilder.havingL}
                            />
                          </div>
                        </Row>
                      </div>
                      <div className="col d-flex w-100 ml-auto">
                        <FormGroup className="align-self-center justify-content-end m-0 ml-auto">
                          <CopyButton
                            id={`copy-btn-${this.props.data.id}`}
                            languageCode={this.props.language.code}
                            handleCopy={this.handleCopy}
                          />
                          <RemoveButton
                            id={`remove-btn-${this.props.data.id}`}
                            languageCode={this.props.language.code}
                            handleRemoveColumn={this.handleRemoveColumn}
                          />
                        </FormGroup>
                      </div>
                    </Row>
                  </Container>
                </Form>
              </CardBody>
            </Card>
          </div>
        )}
      </Draggable>
    );
  }
}

QueryColumn.propTypes = {
  data: PropTypes.shape({
    column_alias: PropTypes.string,
    column_filter: PropTypes.string,
    column_order_dir: PropTypes.bool,
    column_order: PropTypes.bool,
    table_alias: PropTypes.string,
    table_name: PropTypes.string,
    column_name: PropTypes.string,
    id: PropTypes.number,
    display_in_query: PropTypes.bool,
    table_schema: PropTypes.string,
    data_type: PropTypes.string,
    column_distinct_on: PropTypes.bool,
    column_group_by: PropTypes.bool,
    column_aggregate: PropTypes.string,
    subqueryId: PropTypes.number,
    filter_as_having: PropTypes.bool,
  }),
  removeColumn: PropTypes.func,
  addColumn: PropTypes.func,
  updateColumn: PropTypes.func,
  id: PropTypes.string,
  index: PropTypes.number,
  distinct: PropTypes.bool,
  language: PropTypes.shape({ code: PropTypes.string }),
  queries: PropTypes.arrayOf(PropTypes.shape({})),
};

const mapStateToProps = store => ({
  distinct: store.query.distinct,
  language: store.settings.language,
  queries: store.queries.filter(query => query.id !== 0)
    .sort((query1, query2) => query1.id - query2.id),
});

const mapDispatchToProps = {
  updateColumn,
  removeColumn,
  addColumn,
};

export default connect(mapStateToProps, mapDispatchToProps)(QueryColumn);
