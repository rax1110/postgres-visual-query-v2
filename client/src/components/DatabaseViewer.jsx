import React from 'react';
import { connect } from 'react-redux';
import { Scrollbars } from 'react-custom-scrollbars';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import DatabaseTable from './DatabaseTable';
import { filterTable } from '../utils/filterTable';

export const DatabaseViewer = (props) => {
  const constructData = (table) => {
    const data = {
      table_schema: table.table_schema,
      table_name: table.table_name,
      table_type: table.table_type,
      table_alias: '',
    };

    let constraints = JSON.parse(JSON.stringify(props.constraints));

    constraints = constraints.filter(constraint => constraint.table_schema === data.table_schema
      && constraint.table_name === data.table_name);

    let columns = JSON.parse(JSON.stringify(props.columns));

    columns = columns.filter(column => column.table_name === data.table_name
      && column.table_schema === data.table_schema).map((column) => {
      const col = column;

      col.constraints = constraints.filter(
        constraint => _.includes(constraint.column_name, column.column_name),
      );

      delete col.table_name;
      delete col.table_schema;
      return col;
    });

    data.columns = columns;

    return data;
  };

  return (
    <div className="flex-fill">
      <Scrollbars className="d-flex" autoHide>
        <div className="mt-1 pr-2">
          {props.tables.map((table, index) => {
            const checked = props.queryTable.some(
              queryTable => _.isEqual(table.table_name, queryTable.table_name)
                && _.isEqual(table.table_schema, queryTable.table_schema),
            );

            const id = `database-table-${index}`;

            return table.table_schema === props.selectedSchema
              && filterTable(table, props.searchExpr)
              && (
                <DatabaseTable
                  data={constructData(table)}
                  checked={checked}
                  key={id}
                  id={id}
                />
              );
          })}
        </div>
      </Scrollbars>
    </div>
  );
};

DatabaseViewer.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  constraints: PropTypes.arrayOf(PropTypes.shape({})),
  columns: PropTypes.arrayOf(PropTypes.shape({})),
  tables: PropTypes.arrayOf(PropTypes.shape({})),
  queryTable: PropTypes.arrayOf(PropTypes.shape({})),
  selectedSchema: PropTypes.string,
  searchExpr: PropTypes.string,
};

const mapStateToProps = store => ({
  tables: store.database.tables,
  schemas: store.database.schemas,
  selectedSchema: store.database.selectedSchema,
  constraints: store.database.constraints,
  columns: store.database.columns,
  queryTable: store.query.tables,
  searchExpr: store.database.searchExpr,
});

export default connect(mapStateToProps)(DatabaseViewer);
