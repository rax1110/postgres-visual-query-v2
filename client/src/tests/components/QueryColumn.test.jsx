import React from 'react';
import { shallow } from 'enzyme';
import { QueryColumn } from '../../components/QueryColumn';

describe('Component: QueryColumn', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      distinct: false,
      language: { code: 'eng' },
      queries: [{
        id: 1,
        queryName: 'Query 1',
        sql: 'SELECT',
      }],
      updateColumn: jest.fn(),
      removeColumn: jest.fn(),
      addColumn: jest.fn(),
      key: 'query-column-1',
      id: 'query-column-1',
      name: 'query-column-1',
      index: 1,
      data: {
        column_aggregate: '',
        column_alias: '',
        column_distinct_on: false,
        column_filter: '',
        column_group_by: false,
        column_name: 'amet_kood',
        column_order: false,
        column_order_dir: true,
        constraints: [{
          column_name: '{amet_kood}',
          constraint_name: 'pk_amet',
          constraint_type: 'PRIMARY KEY',
          foreign_column_name: null,
          foreign_table_name: null,
          foreign_table_schema: null,
          table_name: 'amet',
          table_schema: 'public',
        }],
        data_type: 'smallint',
        display_in_query: true,
        id: 1,
        ordinal_position: 1,
        table_alias: '',
        table_id: 1,
        table_name: 'amet',
        table_schema: 'public',
        subqueryId: 1,
      },
    };
  });

  test('QueryColumn renders with default props', () => {
    component = shallow(<QueryColumn {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('QueryColumn handleSave calls updateColumn once', () => {
    component = shallow(<QueryColumn {...props} />);
    component.instance().handleSave({
      currentTarget: { name: 'subquerySql' },
      target: { name: 'column_aggregate', value: 'AVG' },
    });

    expect(props.updateColumn.mock.calls.length).toBe(2);
  });

  test('QueryColumn handleRemove calls updateColumn once', () => {
    component = shallow(<QueryColumn {...props} />);

    component.instance().handleRemove({ currentTarget: { name: '' }, target: 'column_alias' });

    expect(props.updateColumn.mock.calls.length).toBe(2);
  });

  test('QueryColumn handleSwitch calls updateColumn once', () => {
    component = shallow(<QueryColumn {...props} />);

    component.instance().handleSwitch({ target: 'column_order' });

    expect(props.updateColumn.mock.calls.length).toBe(2);
  });

  test('QueryColumn handleRemoveColumn calls removeColumn once', () => {
    component = shallow(<QueryColumn {...props} />);

    component.instance().handleRemoveColumn();

    expect(props.removeColumn.mock.calls.length).toBe(1);
  });

  test('QueryColumn handleCopy calls addColumn once', () => {
    component = shallow(<QueryColumn {...props} />);

    component.instance().handleCopy();

    expect(props.addColumn.mock.calls.length).toBe(1);
  });

  test('QueryColumn componentDidMount calls updateColumn once', () => {
    component = shallow(<QueryColumn {...props} />);

    component.instance().componentDidMount();

    expect(props.updateColumn.mock.calls.length).toBe(2);
  });
});
