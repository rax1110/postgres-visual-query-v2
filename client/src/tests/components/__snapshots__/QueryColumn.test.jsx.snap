// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: QueryColumn QueryColumn renders with default props 1`] = `ShallowWrapper {}`;
